Game implemetation

mineserver228 opened the pull request:
- Wrote game 2048 with visualisation in pygame.
- Added controls for play.
- Added score(without saving).
- Added game colouring.

mineserver228 committed:
Added calculating

mineserver228 committed:
Added filling of empty cells after pressing keys.

mineserver228 committed:
Updated gitignore

mineserver228 committed:
Added filling of empty cells after pressing keys.

mineserver228 committed:
Accidentally added file was deleted.

mineserver228 committed:
Added array rendering and the ability to press keys.

mineserver228 committed:
Removed test lines of code, removed two unnecessary libraries.

mineserver228 committed:
Fixed bug with the crash of the game when the entire field is filled.
Added numbers in cells (powers of twos).
Added score.
Added random coloring of cells.

mineserver228 committed:
Fixed bug with the crash of the game when the entire field is filled.
Added numbers in cells (powers of twos).
Added score.
Added random coloring of cells.

Vor-Art commented on .idea/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# Default ignored files
+/shelf/
+/workspace.xml
+/.idea
```
Лучше удалить этот файл, если он добавлен по ошибке

Vor-Art commented on Game_math_functions.py:
```diff
@@ -0,0 +1,111 @@
+colours = []
+
+# filling game desk at the start
+def mass_fill(array):
```
Немного не понятно название функции. Если она заполняет массив вначале, тььогда лучше её назвать `array_init` или `field_init`.

Vor-Art commented on Game_math_functions.py:
```diff
@@ -0,0 +1,111 @@
+
+# filling game desk at the start
+def mass_fill(array):
+    m = []
+    for i in range(x):
+        m.append(0)
+    for i in range(y):
+        array.append(m.copy())
```
Создание массива верное, но для этих целей в питоне есть генераторы.
Можно переписать инициализацию на `array = [[0] * x for i in range(y)]`.

Vor-Art replied:
Также, сейчас есть потанциальный баг, если передать на вход функции не пустой массив, то к нему припишется в конец новые строчки. Я думаю, это не совсем то, что ты хочешь.
Варианты исправления:
1) Не принемать на вход функции ничего.
2) Переписать на гениратор

Vor-Art commented on Game_math_functions.py:
```diff
@@ -0,0 +1,111 @@
+        m.append(0)
+    for i in range(y):
+        array.append(m.copy())
+    if random.randint(0, 1) == 0:
+        array[random.randint(0, x - 1)][random.randint(0, y - 1)] = 4
+    elif random.randint(0, 1) == 0:
+        array[random.randint(0, x - 1)][random.randint(0, y - 1)] = 2
+    else:
+        array[random.randint(0, x - 1)][random.randint(0, y - 1)] = 2
+        array[random.randint(0, x - 1)][random.randint(0, y - 1)] = 2
+    return array
```
Как вариант, от этого кода можно избавиться, если после очищения массива сразу вызывать функцию `add_blocks`.

Vor-Art commented on Game_math_functions.py:
```diff
@@ -0,0 +1,111 @@
+    global score
+    for i in range(x - 1):
+        for g in range(y):
+            if array[g][i + 1] == 0:
+                array[g][i + 1] = array[g][i]
+                array[g][i] = 0
+            elif array[g][i + 1] == array[g][i]:
+                array[g][i + 1] *= 2
+                array[g][i] = 0
+                score += array[g][i + 1]
```
Проблема со сдвигами:
1) есть блоки не сдвигаются до упора на право `0240` -> сейчас `0204`, должно `0024`
2) порядок суммирования одинаковых блоков  `0444` -> сейчас `0084`, должно `0048`

Чтобы исправиить, нужно проверять справа на лево.

Vor-Art replied:
С остальными сдвигами также

Vor-Art commented on Game_math_functions.py:
```diff
@@ -0,0 +1,111 @@
+    for i in range(random.randint(1, 2)):
+        randx = random.randint(0, x - 1)
+        randy = random.randint(0, y - 1)
+        i = 0
```
Почему внутренняя переменная счёта называется также, как и во внешнем цикле?
Это баг.

Vor-Art commented on Game_math_functions.py:
```diff
@@ -0,0 +1,111 @@
+            i += 1
+            randx = random.randint(0, x - 1)
+            randy = random.randint(0, y - 1)
+        if random.randint(0, 1) == 0:
+            array[randy][randx] = 2
+        else:
+            array[randy][randx] = 4
```
Нужно пересмотреть вероятности.
Согласно [wikipedia](https://ru.m.wikipedia.org/wiki/2048_(игра)):
> В каждом раунде появляется плитка номинала «2» (с вероятностью 90 %) или «4» (с вероятностью 10 %)

Vor-Art commented on Pygame visualisation.py:
```diff
@@ -0,0 +1,83 @@
+    screen.fill(background_colour)
+    # keys
+    # key DOWN
+    if keyboard.is_pressed("down"):
+        if key_flagd != 1:
+            key_flagd = 1
+            mass = gm.key_down(mass)
+            mass = gm.add_blocks(mass)
+    if not keyboard.is_pressed("down"):
+        if key_flagd == 1:
+            key_flagd = 0
+    # key UP
+    if keyboard.is_pressed("up"):
+        if key_flagu != 1:
+            key_flagu = 1
+            mass = gm.key_up(mass)
+            mass = gm.add_blocks(mass)
+    if not keyboard.is_pressed("up"):
+        if key_flagu == 1:
+            key_flagu = 0
+    # key LEFT
+    if keyboard.is_pressed("left"):
+        if key_flagl != 1:
+            key_flagl = 1
+            mass = gm.key_left(mass)
+            mass = gm.add_blocks(mass)
+    if not keyboard.is_pressed("left"):
+        if key_flagl == 1:
+            key_flagl = 0
+    # key RIGHT
+    if keyboard.is_pressed("right"):
+        if key_flagr != 1:
+            key_flagr = 1
+            mass = gm.key_right(mass)
+            mass = gm.add_blocks(mass)
+    if not keyboard.is_pressed("right"):
+        if key_flagr == 1:
+            key_flagr = 0
```
Это можно переписать с помощью цикла:
```python3
keys = ["down", "up", "left",  "right"]
actions = [gm.key_down, gm.key_up, gm.key_left, gm.key_right]

for key, action in zip(keys, actions):
   if keyboard.is_pressed(key):
        mass = action(mass)
        mass = gm.add_blocks(mass)
        ...
```

Vor-Art commented on Pygame visualisation.py:
```diff
@@ -0,0 +1,83 @@
+            # colours not working
+            colour = 0
+            division = mass[g][i]
+            while not division < 2:
+                colour += 1
+                division = division // 2
+            pygame.draw.rect(screen,
```
Ты здесь написал фунцию логарифма по основанию 2 :)
Есть готовое решение в одну строчку
```python3
import math

value = mass[g][i]
colour = int(math.log2(value))   # python 3.3 or later
```

Vor-Art commented on Pygame visualisation.py:
```diff
@@ -0,0 +1,83 @@
+            pygame.draw.rect(screen,
+                (gm.colours[colour * 3 + 1], gm.colours[colour * 3 + 2], gm.colours[colour * 3 + 3]),
+                (110 * i, 110 * g, 100, 100))
+            if not mass[g][i] == 0:
```
Не ошибка, просто замечание, что это можно написть более читаемо:
`if mass[g][i] != 0:` или `if mass[g][i] not 0:`

mineserver228 committed:
Added autoclicker function.
Fixed bugs with blocks moving.
Enchanted interface.
Added "Best score"(without saving).
You can stop autoclicker, when you press key "space".

mineserver228 committed:
Reduced keys code
